### ISSUE TITLE
Introduce color-based holdouts in graphs and multigraphs.

### ISSUE THREAD

**Issue by LucaCappelletti94** [LucaCappelletti94/ensmallen_graph#16]
In some experimental setups, we need to have color-based holdouts.

For instance, suppose to have a graph with edges of two colors: red and blue.

In some experimental setup we want all the red edges in the training set alongside with 90% of the blue edges, while the remaining 10% of the blue edges remain in the validation set.

We may be able to do this by introducing a filter method that generates a subgraph based on the color of the edges, that may be called `filter_edges_by_type(self, edge_type)->Graph`. The holdout mechanism might look as follows:

1. Split the graph into `red_graph` and `blue_graph` using the `filter_edges_by_type` method.
2. Randomly split the `blue_graph` into the desired percentages using `random_holdout` into `train_blue_graph and test_blue_graph`. In the expected use case, the blue graph will have a lot of disconnected components while the red graph should be relatively connected, so the number of newly created graph components should be relatively low.
3. Sum the graphs `red_graph + train_blue_graph` to obtain the `train_graph`.

The limitation of this approach is that, if the blue graph connects different components that the red graph does not reach we would be creating new disconnected components in the training graph that are connected in the original graph. This issue might be relatively minimal in our use case and can be measured by using the `report` method to verify if, in our use case it has a significant impact.

@justaddcoffee, @pnrobinson and @vidarmehr do you think this approach should resolve the holdout as discussed in Friday's call on synthetic lethality?

**Comment by vidarmehr** [comment]
@LucaCappelletti94 Thanks, Luca! Yes. I think this could be a right approach. 

> The limitation of this approach is that, if the blue graph connects different components that the red graph does not reach we would be creating new disconnected components in the training graph that are connected in the original graph. This issue might be relatively minimal in our use case and can be measured by using the report method to verify if, in our use case it has a significant impact.

Can we check if removing a blue edge creates a disconnected component or not? If it does not, then we remove the blue edge.

Would that be easy to extend this approach to the case that there are three edge colors(blue, red, green) or more in the graph, assuming that we are still interested in predicting just the blue edges?
Thanks.
